clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -354,7 +317,8 @@ private boolean shouldHandleAuthorizationError(RuntimeException exception) {
         return false;
     }
 
-    private long sendProducerData(long now) {
+    // Visible for testing
+    protected long sendProducerData(long now) {
```

Comment:
The bug could also manifest as a different record batch being sent to the broker. In UnifiedLog.trimInvalidBytes(), we trim extra bytes from the record batch. On the leader side, we don't expect the record batch to have extra bytes. It would be useful to log a warning in this is not the case.